### PR TITLE
Adding Pipeline

### PR DESCRIPTION
**Pull request by filip-halt**
Signed-off-by: Filip Haltmayer <filip.haltmayer@zilliz.com>

### CODE REVIEW

File: towhee/pipeline/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+        self._dag = {}
+        self._input = None
+        self._output = None
+
+    def set_config(self, config) -> 'Pipeline':
+        self._config = config
+
+    @classmethod
+    def input(cls, output_schema) -> 'Pipeline':
+        uid = uuid.uuid4().hex
```
Comment (junjiejiangjjj) [inline towhee/pipeline/pipeline.py:37]: why not give a special name, such as `_ start`, then we can find the start node easily

File: towhee/pipeline/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+            'fn': None,
+            'input': None,
+            'output': output_schema,
+            'fn_type': 'input',
+            'iteration': 'map'
+        }
+        return pipeline
+
+    def output(self, input_schema, engine = 'local') -> Callable:
+        uid = uuid.uuid4().hex
```
Comment (junjiejiangjjj) [inline towhee/pipeline/pipeline.py:50]: The same as the comment of `input`

File: towhee/pipeline/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+            'fn_type': 'output',
+            'iteration': 'map'
+        }
+        #TODO Return the DC once dc is decided.
+        return self
+
+
+    def map(self, input_schema, output_schema, fn, config = None) -> 'Pipeline':
+        uid = uuid.uuid4().hex
+        if isinstance(fn, tuple):
```
Comment (junjiejiangjjj) [inline towhee/pipeline/pipeline.py:66]: Why the instance is tuple

Comment (junjiejiangjjj) [reply towhee/pipeline/pipeline.py:66]: the if-else occurs in many places

File: towhee/pipeline/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+            'fn': fn,
+            'input': input_schema,
+            'output': output_schema,
+            'fn_type': fn_type,
+            'iteration': 'map',
+            'config': config
+        }
+        return self
+
+    def flat_map(self, filter_schema, output_schema, fn, config = None) -> 'Pipeline':
```
Comment (junjiejiangjjj) [inline towhee/pipeline/pipeline.py:83]: Why is `filter_schema`, not `input_schema`

File: towhee/pipeline/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
+from typing import Callable
+from towhee.pipeline.action import Action
+import uuid
+import json
+
+# pylint: disable=protected-access
+# pylint: disable=redefined-builtin
+
+
+class Pipeline:
```
Comment (shiyu22) [inline towhee/pipeline/pipeline.py:10]: I think we need to add `tag` key to the dag, which defaults to 'main'.

File: towhee/pipeline/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+        """
+        uid = uuid.uuid4().hex
+        fn_action = self._to_action(fn)
+        fn_type, fn = fn_action.serialize()
+
+        self._dag[uid] = {
+            'fn': fn,
+            'input': input_schema,
+            'output': output_schema,
+            'filter_type': fn_type,
```
Comment (shiyu22) [inline towhee/pipeline/pipeline.py:172]: Why `filter_type` and not `fn_type`?

File: towhee/pipeline/pipeline.py (hunk 5 above)
Comment (filip-halt) [reply towhee/pipeline/pipeline.py:10]: The issue came up of how does the user input that data. Ops are sepearte from pipeline. And currently ops are created with ops.repo.name(args, kwargs). One option is to do ops.repo.name['tag'](args, kwargs), but that is pretty ugly.

File: towhee/pipeline/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        }
+        #TODO Return the DC once dc is decided.
+        return self
+
+
+    def map(self, input_schema, output_schema, fn, config = None) -> 'Pipeline':
+        """One to one map of function on inputs.
+
+        Args:
+            input_schema (str | tuple): The column/s to perform map on.
```
Comment (shiyu22) [inline towhee/pipeline/pipeline.py:102]: Can you set the type of input_schema and output_schema to Tuple? same as other type of nodes.

**Pull request closed without merging**